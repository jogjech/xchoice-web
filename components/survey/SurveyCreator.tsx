import React, { FunctionComponent, useState } from "react";
import { Form, Input, Button, Result, Alert } from "antd";
import QuestionCreator from "./question/QuestionCreator";
import { postSurvey } from "../../apis/survey";
import { useAuth0 } from "@auth0/auth0-react";
import { useRouter } from "next/router";

interface Props {}

const SurveyCreator: FunctionComponent<Props> = () => {
  const { getAccessTokenSilently } = useAuth0();

  const [form] = Form.useForm();
  const [posting, setPosting] = useState(false);
  const [postError, setPostError] = useState(undefined);
  const [posted, setPosted] = useState(false);
  const [surveyId, setSurveyId] = useState("");
  const [redirecting, setRedirecting] = useState(false);
  const router = useRouter();

  const handleFormFinish = async (formData: {
    surveyTitle: string;
    questions: [
      {
        questionTitle: string;
        choices: [{ text: string }];
      }
    ];
  }) => {
    console.log(formData);
    setPosting(true);
    setPostError(undefined);

    const accessToken = await getAccessTokenSilently();
    const result = await postSurvey(formData, accessToken);

    setPostError(result.error);
    setPosted(!result.isError);
    setSurveyId(result.surveyId);
    setPosting(false);
  };

  return (
    <>
      {posted ? (
        <Result
          status="success"
          title="Your survey is posted!"
          subTitle={`Survey ID: ${surveyId}.`}
          extra={[
            <Button
              type="primary"
              key="console"
              loading={redirecting}
              onClick={() => {
                setRedirecting(true);
                router.push(`/survey/${surveyId}`);
              }}
            >
              View my survey
            </Button>,
          ]}
        />
      ) : (
        <>
          <div>
            <p className="title">Create your survey</p>
          </div>
          <Form
            layout="vertical"
            form={form}
            initialValues={{
              questions: [
                {
                  choices: [
                    {
                      text: "",
                    },
                  ],
                },
              ],
            }}
            onFinish={handleFormFinish}
            autoComplete="off"
          >
            <p className="sub-title">Basic Info</p>
            <Form.Item
              label="Survey Title"
              name="surveyTitle"
              rules={[{ required: true, message: "Missing survey title" }]}
            >
              <Input placeholder="Survey Title"></Input>
            </Form.Item>
            <QuestionCreator></QuestionCreator>
            <Form.Item>
              <Button type="primary" htmlType="submit" loading={posting}>
                Submit
              </Button>
            </Form.Item>
          </Form>
          {postError && (
            <Alert
              message="Failed to create survey. Please try again."
              type="error"
            ></Alert>
          )}
        </>
      )}
    </>
  );
};

export default SurveyCreator;
