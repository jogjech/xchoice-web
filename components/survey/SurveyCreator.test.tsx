import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import SurveyCreator from "./SurveyCreator";

describe("Survey creator", () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<SurveyCreator></SurveyCreator>);
  });

  it("renders correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("contains form", () => {
    const form = wrapper.find("ForwardRef(InternalForm)");
    expect(form.exists()).toBeTruthy();

    const initialValues = form.prop("initialValues");
    expect(initialValues).toEqual({
      questions: [
        {
          choices: [
            {
              text: "",
            },
          ],
        },
      ],
    });

    expect(form.find(".sub-title").text()).toBe("Basic Info");

    expect(form.find("QuestionCreator").exists()).toBeTruthy();

    expect(form.find("Button").text()).toBe("Submit");
  });
});
