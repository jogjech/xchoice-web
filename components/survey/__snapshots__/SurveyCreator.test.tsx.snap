// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Survey creator renders correctly 1`] = `
<Fragment>
  <div>
    <p
      className="title"
    >
      Create your survey
    </p>
  </div>
  <ForwardRef(InternalForm)
    autoComplete="off"
    form={
      Object {
        "__INTERNAL__": Object {
          "itemRef": [Function],
        },
        "getFieldError": [Function],
        "getFieldInstance": [Function],
        "getFieldValue": [Function],
        "getFieldsError": [Function],
        "getFieldsValue": [Function],
        "getInternalHooks": [Function],
        "isFieldTouched": [Function],
        "isFieldValidating": [Function],
        "isFieldsTouched": [Function],
        "isFieldsValidating": [Function],
        "resetFields": [Function],
        "scrollToField": [Function],
        "setFields": [Function],
        "setFieldsValue": [Function],
        "submit": [Function],
        "validateFields": [Function],
      }
    }
    initialValues={
      Object {
        "questions": Array [
          Object {
            "choices": Array [
              Object {
                "text": "",
              },
            ],
          },
        ],
      }
    }
    layout="vertical"
    onFinish={[Function]}
  >
    <p
      className="sub-title"
    >
      Basic Info
    </p>
    <FormItem
      label="Survey Title"
      name="surveyTitle"
      rules={
        Array [
          Object {
            "message": "Missing survey title",
            "required": true,
          },
        ]
      }
    >
      <Input
        maxLength={100}
        placeholder="Survey Title"
        type="text"
      />
    </FormItem>
    <QuestionCreator />
    <FormItem>
      <Button
        block={false}
        ghost={false}
        htmlType="submit"
        loading={false}
        type="primary"
      >
        Submit
      </Button>
    </FormItem>
  </ForwardRef(InternalForm)>
</Fragment>
`;
